### PR TITLE
Use the more accurate distance routines in geographiclib, if available.

### PR DESCRIPTION
**Pull request by cffk**
This follows up on an E-mail thread on geopy@googlegroups.com from
September 2011, "Patch to provide GeodesicDistance for geopy".  It
provides a more accurate solution for geodesics compared to Vincenty by
using the python package geographiclib if it is available.  In addition
the distance calculation always succeeds.  It's possible that it
addresses issue #140 also.  Still to do: add tests (but I haven't
figured out the testing framework yet).

### CODE REVIEW

**Comment by cffk** [comment]
This patch will also fix issue #30, I believe.

**Comment by ijl** [comment]
@cffk, cool, can you perhaps refactor `geopy.distance` to be a directory, with the python implementation in one module, the `if HAVE_GEOGRAPHICLIB:` branch in another, and the init choosing? This would also need regression tests run against each module.

**Comment by cffk** [comment]
Hmm, it's not so obvious how to do this...  This issues are:

* distance.py includes a routine for the great circle distance and this
  needs to remain regardless of the choice of how to do geodesics.

* The present names for the ellipsoid distance are VincentyDistance and
  vincenty.  But these are inappropriate names for the more accurate
  method.

Here is a possible way of dealing with the situation.

* If HAVE_GEOGRAPHICLIB is true, then the code works as my pull request.

* If it is false, then GeodesicDistance and geodesic are set to vincenty
  (and the VincentyDistance and vincenty names are retained).

* In either case, the names VincentyDistance and vincenty are marked as
  deprecated.  People should be encouraged to use the name
  GeodesicDistance.

* After some suitable interval, the either/or implementation can be
  substituted (i.e., if HAVE_GEOGRAPHICLIB, then the vincenty method
  won't even be defined; otherwise vincenty method will be defined but
  will be called geodesic and not vincenty).

  --Charles

**Comment by ijl** [comment]
That makes sense to me.

**Comment by cffk** [comment]
OK, I'll make this change this weekend.

**Comment by micahcochran** [comment]
I like the idea of the patch.    Shouldn't this patch atleast have a vincenty class that issues a [deprecation warning](https://docs.python.org/2/library/warnings.html), so as to not entirely break the API?  

That could just be a dummy class that with and __init__ function issues the DeprecationWarning, that is derived from geodesic.

Something like this (this is pseudocode):
```Python
import warnings

class vincenty(geodesic):
    def __init__(self, *args, **kwargs):
        warnings.warn('Replace class distance.vincenty with class geodesic', DeprecationWarning)
        super(geodesic, self).__init__(*args, **kwargs)
```

**Comment by cffk** [comment]
This is a significant redo of the original pull request.

Originally, vincenty was replaced by the geodesic method.  Now both are
provided.  In addition, the geodesic method will use pyproj if it's
available in preference to geographiclib.  Both pyproj and geographiclib
implement exactly the same algorithm.  However, pyproj's implementation
is in compiled C code and so faster.

In the process of testing, I needed to scoop up a couple of pending pull
requests: #187 and #194.  This patch fixes issues: #30 #186 #242.

Here is some timing data.  Times are in microsecs per call

~~~
method                          dist    dest
geopy great_circle              20.4    17.1
geopy vincenty                  40.3    30.4
goopy geodesic(pyproj)          37.1    31.1
geopy geodesic(geographiclib)  302.9   124.1
geographiclib                  279.1   104.6
pyproj                          17.6    15.7
pyproj vector call               2.8     1.4
~~~

Here is the accuracy of the geodesic calculations based on the Geodesic
Test Set

https://doi.org/10.5281/zenodo.32156

The errors are given in units of micons (1e-6 m)

~~~                            
method                        distance destination
geopy vincenty                 205.629  141.945
goopy geodesic(pyproj)           0.007    0.013
geopy geodesic(geographiclib)    0.011    0.010
~~~

Without hannosche's pull request #194, the error in the destination
calculation for vincenty is 8.98 meters.

19.2% of the tests cases failed with vincenty.distance (iterations =
20).  However the test set is skewed towards cases which would case this
failure.

With random points on the WGS84 ellipsoid, the Vincenty algorithm is
guaranteed to fail 16.6 out of 1000000 times (the correct solution is an
unstable fixed point of the Vincenty method).

With the geopy implementation of Vincenty and iterations = 20, the
failure rate is 82.8 per 1000000.  With iterations = 200, the failure
rate is 21.2 per 1000000.

Even though these rates are small, failures can be quite common.  For
example in a dataset of 1000 random points (think the worlds airports,
perhaps), computing the full distance matrix would fail on average 16
times (with iterations = 20).

**Comment by flaviotruzzi** [comment]
@cffk Do you know if this is going to be merged some day?

**Comment by cffk** [comment]
I haven't had any feedback from the package maintainers since I made the changes in May.  Perhaps @micahcochran can weigh in?

**Comment by micahcochran** [comment]
@cffk I have no objections to this being merged. I did not mean for my critique to significantly slow down the process of getting this merged.  I don't have any information about when/if this will be merged.

**Comment by cffk** [comment]
@flaviotruzzi I just the next step for you to take is to ping the gatekeepers of geopy about
merging this pull request.

**Comment by flaviotruzzi** [comment]
@cffk I am a user of geopy, but I am not familiar with the procedures for merging here, is there a main maintainer that we should ping?

**Comment by cffk** [comment]
I looked at the major contributors.  So perhaps @ijl or @jqnatividad  can be persuaded to comment on the pull request and merge if appropriate.

**Comment by morsedl** [comment]
I've not yet used GeoPy, but soon need to use it or something else, Distance precision is very important for my project, so I probably can't use GeoPy until this branch is merged.

So, +1 for merging soon! :-)

**Comment by micahcochran** [comment]
@morsedl If you need precision, I'd suggest looking into the python version of [GeographicLib](https://pypi.python.org/pypi/geographiclib).  It is the work of @cffk 

Its [documentation](https://geographiclib.sourceforge.io/1.49/python/index.html) has some [examples](https://geographiclib.sourceforge.io/1.49/python/examples.html).

It is available right now and has had a release in the past couple month.

File: geopy/distance.py [hunk 1]
```diff
@@ -15,13 +15,29 @@
 :const:`distance.EARTH_RADIUS`, so it can be customized (it should
 always be in kilometers, however).
 
-Vincenty distance (:class:`.vincenty`) uses a more accurate ellipsoidal model
```
Comment (KostyaEsmukov) [inline geopy/distance.py]: But Vincenty is still there, right? And it's the default `distance` when geodesic is not available (a pretty common use-case, I believe). So I suggest to return back this part of doc.

File: geopy/distance.py [hunk 2]
```diff
@@ -35,40 +51,47 @@
...
-    538.3904451566326
+    >>> print(distance.distance(newport_ri, cleveland_oh).miles)
+    538.390445368
+
+    >>> wellington = (-41.32, 174.81)
+    >>> salamanca = (40.96, -5.50)
+    >>> print(distance.distance(wellington, salamanca).km)
+    19959.6792674
+
+The second example above fails unless geographiclib has been installed.
```
Comment (KostyaEsmukov) [inline geopy/distance.py]: By `fails` you mean that there's a significant error, right? Maybe you could show a difference by also calling `print(distance.vincenty(wellington, salamanca).km)` there?

File: geopy/distance.py [hunk 3]
```diff
@@ -15,13 +15,29 @@
...
+(:class:`.vincenty`) which is accurate to 0.2 mm and the distance
+calculation fails to converge for nearly antipodal points.
+
+If either of the python packages pyproj or geographiclib is available,
+then the method given by
+[Karney (2013)](https://doi.org/10.1007%2Fs00190-012-0578-z)
+(:class:`.geodesic`) is also provided.  This is accurate to round-off
+and always converges.  You can install pyproj and geographiclib with
+
+    pip install pyproj geographiclib
```
Comment (KostyaEsmukov) [inline geopy/distance.py]: Is there a reason to need to have both installed? From what I see here either one of them should be enough.

File: geopy/distance.py [hunk 4]
```diff
@@ -79,6 +102,17 @@
 from geopy.point import Point
 from geopy.compat import string_compare
 
+HAVE_PROJ = HAVE_GEODESIC = False
+try:
+    from pyproj import Geod
+    HAVE_PROJ = HAVE_GEODESIC = True
```
Comment (KostyaEsmukov) [inline geopy/distance.py]: I personally don't like having booleans there. Take this line for example: `geographiclib` presence isn't even checked, though the boolean claims that it's there.

I'm good with booleans if these imports are tested individually, but in this case we would have a redundant import of `geographiclib` when `pyproj` is installed.

Maybe using string is a better choice there? Like this:
```
try:
    from pyproj import Geod
    GEODESIC_METHOD = 'pyproj'
except ImportError:
    try:
        from geographiclib.geodesic import Geodesic
        GEODESIC_METHOD = 'geographiclib'
    except ImportError:
        GEODESIC_METHOD = None  # not available
```

File: geopy/distance.py [hunk 5]
```diff
@@ -284,6 +316,8 @@ def destination(self, point, bearing, distance=None): # pylint: disable=W0621
 
         return Point(units.degrees(radians=lat2), units.degrees(radians=lng2))
 
+GreatCircleDistance = great_circle
```
Comment (KostyaEsmukov) [inline geopy/distance.py]: This line seems to be redundant: there's the same one at the bottom of this module.

File: geopy/distance.py [hunk 6]
```diff
@@ -526,8 +561,118 @@ def destination(self, point, bearing, distance=None): # pylint: disable=W0621
...
+    """
 
-distance = VincentyDistance = vincenty
+    ellipsoid_key = None
+    ELLIPSOID = None
+    geod = None
+
+    def __init__(self, *args, **kwargs):
+        if not HAVE_GEODESIC:
+            raise NotImplementedError(
```
Comment (KostyaEsmukov) [inline geopy/distance.py]: I think it should be `ImportError`.

File: geopy/distance.py [hunk 7]
```diff
@@ -526,8 +561,118 @@ def destination(self, point, bearing, distance=None): # pylint: disable=W0621
...
+                "neither pyproj nor geographiclib is available")
+        self.set_ellipsoid(kwargs.pop('ellipsoid', 'WGS-84'))
+        self.method = "pyproj" if HAVE_PROJ else "geographiclib"
```
Comment (KostyaEsmukov) [inline geopy/distance.py]: I would suggest to make 2 different implementations of `geodesic` class instead (probably having a common base class): one for `pyproj` and another for `geographiclib`. This way we will get rid of these ugly `if`s below and we will be able to run tests against both implementations (especially `geographiclib` when having `pyproj` installed).

File: test/test_distance.py [hunk 8]
```diff
@@ -206,3 +217,93 @@ def test_should_get_distinct_results_for_different_ellipsoids(self):
                        for y in range(len(results))
                        if x != y)
 
+if HAVE_GEODESIC:
```
Comment (KostyaEsmukov) [inline test/test_distance.py]: Please use unittest decorator instead:

```
@unittest.skipUnless(  # pylint: disable=R0904,C0111
    HAVE_GEODESIC,
    "Neither pyproj nor geographiclib is installed"
)
class TestWhenComputingGeodesicDistance(CommonDistanceCases):
    ...
```

**Comment by cffk** [comment]
@KostyaEsmukov Before implementing your suggestions, it might be worth stepping back to see if there's a simpler approach...

The problem arises because the Vincenty method fails to converge for nearly antipodal points.  For a long time, this was excused because computing such distances was rare.  However, it has real negative consequences.  For example, computing the distance matrix for 1000 randomly selected points results in about 8 failures of the Vincenty algorithm.  This means that the triangle inequality is not satisfied which, in turn, precludes the use of algorithms such as the VP tree for computing, e.g., the nearest port for a ship at sea.  See my [answer to a question on stackoverflow](https://stackoverflow.com/questions/45440700/calculate-distance-to-shore-or-coastline-for-a-vessel/45480555#45480555).

In 2009 I provided improved algorithms for computing geodesics.  This is more accurate (1e-8 m instead of 1e-4 m), but, more importantly, the distance calculation always converges.  The initial implementation was in C++; however, I provided a Python implementation in 2011.  This pull request provides hooks to this implementation.

Unfortunately, along the way the pull request became successively more convoluted:

* it retains the Vincenty algorithm;
* my native Python implementation is used if available;
* except that, if pyproj is available, that is used instead.

(pyproj uses a C implementation of my algorithms.)  This plethora of choices is pretty much guaranteed to cause confusion on the part of end users.

Here's my recommendation:

Replace Vincenty with my algorithms.  Either make geographiclib a dependency of geopy or just include the code as part of geopy.  (This approach has worked well for proj.4 which includes the C implementation. It helps that the source files are kept consistent so that changes in one library can be quickly brought into the other.)

This would also allow immediate access to routines for computing the area of geodesic polygons and for computing differential properties of geodesics (e.g., how much does the end point of a geodesic point move when the initial azimuth is changed by a small amount) which are useful, e.g., in doing adjustments of triangulation networks.

The potential push back is that my algorithms are 7x (resp. 4x) slower than the Vincenty distance (resp. destination).  I would say "ignore this".  Most casual users won't care (the distance calculation takes 0.3 ms).  Hard-core users can now use much better algorithms for distances involving many points because the triangle inequality is satisfied.

In this scheme, the routines with "vincenty" in their names would be deprecated and would merely call the corresponding routines with "vincenty" replaced by "geodesic".

Various options are available if speed really is an issue:

* hook up to pyproj
* provide a numpy implementation
* use Cython to call the C++ library

However, I regard this as secondary to getting decent algorithms into the hands of your users.

**Comment by KostyaEsmukov** [comment]
Thank you, Charles, for such a comprehensive explanation.

Let me answer each point in turn:

> Replace Vincenty with my algorithms. [...] In this scheme, the routines with "vincenty" in their names would be deprecated and would merely call the corresponding routines with "vincenty" replaced by "geodesic".

Vincenty deprecation sounds reasonable to me, but I would rather not implicitly substitute `vincency` implementation with `geodesic`, because some users might rely on vincenty's inaccuracy or the 4-7x faster computation speed. So I would prefer to still keep vincenty, but issue a warning for its usages with an advice to use a generic `distance` class instead, which provides an up-to-date method of distance computation (`geodesic`, for the moment).

In this case removing vincenty example and explanation from the docs is fine with me, but `vincenty` class should be marked as deprecated with:
- a `warning` in `measure` and `destination` methods with [`DeprecationWarning` category](https://docs.python.org/2/library/warnings.html#warning-categories) and an advice to use `geodesic` instead
- a `.. warning::` block in `vincenty` doc, explaining why `vincenty` is deprecated (because it's less accurate and in some cases fails to converge) and what to use instead (`distance` class).

> Either make geographiclib a dependency of geopy or just include the code as part of geopy.

I'm fine with adding a dependency on `geographiclib`. I would say vendoring is not needed here (unless there's an issue with any of the geopy's supported Python versions which cannot be fixed on your (`geographiclib`) side, but this looks unlikely).

> This would also allow immediate access to routines for computing the area of geodesic polygons and for computing differential properties of geodesics (e.g., how much does the end point of a geodesic point move when the initial azimuth is changed by a small amount) which are useful, e.g., in doing adjustments of triangulation networks.

I would expect that users who actually need that kind of advanced measurements are aware of `geographiclib` itself, and I tend to think that the higher level geopy is probably not for them. Thereby, I'd rather keep the interfaces as simple as they're now, without bringing in peculiar features which aren't needed by most users.

> Various options are available if speed really is an issue: [...]. However, I regard this as secondary to getting decent algorithms into the hands of your users.

I agree with the last point, whereas I believe that speed is *really* an issue for some of the users, so having an easy way (like installing pyproj) to quickly enhance performance without tinkering with binary extensions on the user side would be great. If we tend to think of pyproj as a solution to this, then we may simply add an [extras to `setup.py`](https://github.com/geopy/geopy/blob/67a18801cc87eec40e1f1e8edaa6f297228be91e/setup.py#L32), which would allow to bring in pyproj with simply doing something like `pip install geopy[fastgeodesic]`.

> * hook up to pyproj
> * provide a numpy implementation
> * use Cython to call the C++ library

I would rather not introduce any required binary extension (or required dependency with one, like numpy or pyproj) which might make installation more complicated even if a user is going to use only the geocoders part of the library, so building a native extension is kinda useless for them. Also compiling CPython extensions for pypy (which geopy supports at the moment) might be tricky. Thus I believe that a python-only solution is a requirement here.

To sum up, here are my suggestions:
- vincenty should be deprecated, but not substituted with `geodesic`
- a python-only solution for `geodesic` should be used by default to avoid platform-specific issues
- having a way to optionally install a binary dependency to speed up computations would be great (but I'm totally OK to move this out of the scope of this PR to simplify things for now).

**Comment by cffk** [comment]
I'm not sure how to let the test machines see geographiclib...?

File: geopy/distance.py [hunk 9]
```diff
@@ -15,13 +15,24 @@
...
-of the earth. This is the default distance formula, and is thus aliased as
-``distance.distance``. There are multiple popular ellipsoidal models, and
-which one will be the most accurate depends on where your points are located
-on the earth. The default is the WGS-84 ellipsoid, which is the most globally
-accurate. geopy includes a few other
-models in the distance.ELLIPSOIDS dictionary::
+The geodesic distance is the shortest distance on the surface of an
+ellipsoidal model of the earth.  There are two algorithms used to
+compute this distance.  The default algorithm uses the method is given
+by [Karney (2013)](https://doi.org/10.1007%2Fs00190-012-0578-z)
```
[the remark this answers is not in the document]
Comment (cffk) [reply geopy/distance.py]: Will fix

File: setup.py [hunk 10]
```diff
@@ -6,8 +6,11 @@
 from setuptools import setup, find_packages
 from geopy import __version__ as version
 
-INSTALL_REQUIRES = []
+INSTALL_REQUIRES = [
+    'geographiclib'
```
[the remark this answers is not in the document]
Comment (cffk) [reply setup.py]: Will fix

File: test/test_distance.py [hunk 11]
```diff
@@ -10,6 +10,8 @@
 from geopy.distance import (Distance,
                             GreatCircleDistance,
                             VincentyDistance,
+                            GeodesicDistance,
+                            HAVE_GEODESIC,
```
[the remark this answers is not in the document]
Comment (cffk) [reply test/test_distance.py]: Will fix

File: test/test_distance.py [hunk 12]
```diff
@@ -286,3 +297,91 @@ def test_should_get_distinct_results_for_different_ellipsoids(self):
                        for y in range(len(results))
                        if x != y)
 
+class TestWhenComputingGeodesicDistance(CommonDistanceCases):
```
[the remark this answers is not in the document]
Comment (cffk) [reply test/test_distance.py]: Will fix

File: test/test_distance.py [hunk 13]
```diff
@@ -286,3 +297,91 @@ def test_should_get_distinct_results_for_different_ellipsoids(self):
...
+
+    cls = GeodesicDistance
+
+    def setup(self):
+        self.original_ellipsoid = self.cls.ELLIPSOID
+
+    def teardown(self):
```
[the remark this answers is not in the document]
Comment (cffk) [reply test/test_distance.py]: Will fix

File: test/test_distance.py [hunk 14]
```diff
@@ -286,3 +297,91 @@ def test_should_get_distinct_results_for_different_ellipsoids(self):
...
+           11911190.819018408],
+          [-29.47124, 95.14681, -163.779130441688382,
+           -27.46601, -69.15955, -15.909335945554969,
+           13487015.8381145492]]
+        d = self.cls(ellipsoid = 'WGS-84')
+        km = 1000
+        for l in testcases:
+            (lat1, lon1, azi1, lat2, lon2, azi2, s12) = l
+            # lon2 is the UNROLLed longitude, so normalize
+            lon2 = (lon2 + 360 if lon2 < -180 else
```
[the remark this answers is not in the document]
Comment (cffk) [reply test/test_distance.py]: Will fix (normalization isn't necessary)

File: geopy/distance.py [hunk 15]
```diff
@@ -303,6 +320,97 @@ def destination(self, point, bearing, distance=None): # pylint: disable=W0621
...
+    more information.
+
+    Example::
+
+        >>> from geopy.distance import geodesic
+        >>> newport_ri = (41.49008, -71.312796)
+        >>> cleveland_oh = (41.499498, -81.695391)
+        >>> print(geodesic(newport_ri, cleveland_oh).miles)
+        538.390445368
+    """
```
[the remark this answers is not in the document]
Comment (cffk) [reply geopy/distance.py]: Will fix.  Note that I need to use

    PYTHONPATH=.. make html

File: test/test_distance.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (cffk) [reply test/test_distance.py]: No, I *do* need to normalize.  I retain my original idiom since this puts the longitude in (-180,180] whereas Point's normalization uses [-180,180].  Almost certainly, either will work.

[the remark this answers is not in the document]
Comment (cffk) [reply test/test_distance.py]: I converted the (lat,lon) pairs to Points and the issue disappears.

File: geopy/distance.py [hunk 16]
```diff
@@ -2,26 +2,39 @@
...
 
-Vincenty distance (:class:`.vincenty`) uses a more accurate ellipsoidal model
-of the earth. This is the default distance formula, and is thus aliased as
-``distance.distance``. There are multiple popular ellipsoidal models, and
-which one will be the most accurate depends on where your points are located
-on the earth. The default is the WGS-84 ellipsoid, which is the most globally
-accurate. geopy includes a few other
-models in the distance.ELLIPSOIDS dictionary::
+The geodesic distance is the shortest distance on the surface of an
+ellipsoidal model of the earth.  There are two algorithms used to
```
Comment (KostyaEsmukov) [inline geopy/distance.py]: > There are two algorithms used to compute this distance.

Aren't there three now? Geodesic, Vincenty and GreatCircle.

GreatCircle is not even mentioned in this introduction, so maybe it makes sense to rephrase this sentence to not include a count of the algorithms.

**Comment by cffk** [comment]
OK, I documented that vincenty is deprecated.  I don't issue a warning when vincenty is used -- I think this should wait for a subsequent release?

**Comment by KostyaEsmukov** [comment]
> I think this should wait for a subsequent release?

OK, maybe you're right. There might be some reasons to prefer vincenty over geodesic at the moment (like performance or maybe some yet unknown issues with geodesic).
